Allow setFromIK() with multiple poses to single IK solver

davetcoleman opened the pull request:
Allows a call to setFromIK() that has multiple tips and desired poses to send them all to the same IK solver, if that solver allows it. If no solver exists for the current planning group or that solver does not support non-chain kinematics, the previous MoveIt! behavior is maintained whereby it looks for subgroups that are chains that have their own IK solvers and calls them individually.

This is a follow up of https://github.com/ros-planning/moveit_core/pull/149
### Minor Changes

Reduced duplicate code by moving two common code snippets into their own functions:
- Created new helper function for changing a pose to the base frame of an IK solver
- Created new helper function for converting a Eigen pose to a geometry_msgs pose

Contains outstanding PR:
https://github.com/ros-planning/moveit_core/pull/177

davetcoleman committed:
Allowed setFromIK to do whole body IK solving with multiple tips

davetcoleman committed:
Add check for correct solver type

davetcoleman committed:
Split setToIKSolverFrame() into two functions

davetcoleman committed:
Added new setToRandomPositions function that allows custom random number generator to be specified

davetcoleman committed:
Maintain ordering of poses in order that IK solver expects

davetcoleman committed:
Removed debugging code

davetcoleman committed:
Reduced unnecessary joint position copying

davetcoleman committed:
Addressed consistency_limits issue

davetcoleman committed:
Merge branch 'hydro-devel' into set_from_ik_tips

davetcoleman committed:
Removed duplicate getSubgroups() function, added prefixes for changed log messages

davetcoleman committed:
Bug fix

isucan commented:
@sachinchitta ? I am fine with merging this.

sachinchitta commented on robot_state/include/moveit/robot_state/robot_state.h:
```diff
@@ -611,6 +611,29 @@ as the new values that correspond to the group */
+   * @param pose - input
+   * @param pose_msg - output
+   */
+  void poseToMsg(const Eigen::Affine3d &pose, geometry_msgs::Pose &pose_msg);
```
There's a eigen to msg conversions package that should be used - this is a duplication and should not be in robot_state.h

davetcoleman replied:
That's a really good point. Before this PR this same code was hard-coded into robot_state inline with the code, and I didn't think about this duplication when I created this function. Removed.

sachinchitta commented on robot_state/include/moveit/robot_state/robot_state.h:
```diff
@@ -611,6 +611,29 @@ as the new values that correspond to the group */
+   * @param ik_frame - the name of frame of reference of base of ik solver
+   * @return true if no error
+   */
+  bool setToIKSolverFrame(Eigen::Affine3d &pose, const std::string &ik_frame);
```
I believe that this is already done by the transforms class (transforming a pose to a new frame). Why duplicate it here (unless its not easily available here)?

davetcoleman replied:
At the heart of this function yes, it does call `getGlobalLinkTransform()`. But it encapsulates several other lines of logic that I saw repeated throughout the `setFromIK()` functions that I thought would increase clarity by removing this sub function, including
- Checking if the two frames are already the same
- Checking if the frame is prepended with a '/' or not
- Checking if the LinkModel pointer is valid
- A line of math
- Getting the base frame from the solver

I think it should stay

sachinchitta replied:
Given this this uses a string frame name, it could be used to transform to any frame (not just an ik frame). Calling this function transformPoseToFrame matches better with the syntax in other places. Also, if this function is mostly used internally, maybe it should be private?

davetcoleman replied:
I like `transformPoseToFrame()`, ok. And making it private makes sense.

davetcoleman replied:
Actually, I was making the modifications just now, and I realized this function is really intended just for IK solvers. It first checks if the IK solver frame is the same as the robot_model frame, and if it is it skips the transform. Only if they differ does it change the frame.

If you are worried about adding too many functions, I could remove the first `setToIKSolverFrame` that takes as input the solver pointer, it isn't very useful.

sachinchitta commented on robot_state/src/robot_state.cpp:
```diff
@@ -1231,146 +1225,270 @@ bool moveit::core::RobotState::setFromIK(const JointModelGroup *jmg, const Eigen
+    std::string error_msg;
+    if(!solver->supportsGroup(jmg, &error_msg))
     {
-      if (hasAttachedBody(tip))
```
Its a little hard for me to parse through these changes but this looks like support for attached bodies IK is being taken out?

davetcoleman replied:
Github just did a bad job of overlaying the changes I'm afraid. Both `setFromIK()` and the new `setFromIKSubgroups()` still have the `hasAttachedBody(tip)` logic, it just moved a little further down.

sachinchitta commented on robot_state/src/robot_state.cpp:
```diff
@@ -1402,45 +1522,32 @@ bool moveit::core::RobotState::setFromIK(const JointModelGroup *jmg, const Eigen
 
-bool moveit::core::RobotState::setFromIK(const JointModelGroup *jmg, const EigenSTL::vector_Affine3d &poses_in, const std::vector<std::string> &tips_in,
-                                         unsigned int attempts, double timeout,
-                                         const GroupStateValidityCallbackFn &constraint, const kinematics::KinematicsQueryOptions &options)
```
Is this function being taken out of the API? Why not just add a new function and leave the old ones in to minimize API changes?

davetcoleman replied:
No, it is still included but appears moved, as I noted above.

davetcoleman commented on robot_state/src/robot_state.cpp:
```diff
@@ -1231,146 +1225,270 @@ bool moveit::core::RobotState::setFromIK(const JointModelGroup *jmg, const Eigen
-        if (!lm)
-          return false;
-        current_pose = getGlobalLinkTransform(lm).inverse() * current_pose;
+        if (hasAttachedBody(tip))
```
This is the new location of `hasAttachedBody`

davetcoleman committed:
Removed duplicate functionality poseToMsg function

davetcoleman commented on robot_state/src/robot_state.cpp:
```diff
@@ -1231,146 +1225,270 @@ bool moveit::core::RobotState::setFromIK(const JointModelGroup *jmg, const Eigen
+bool moveit::core::RobotState::setFromIK(const JointModelGroup *jmg, const EigenSTL::vector_Affine3d &poses_in, const std::vector<std::string> &tips_in,
+                                         unsigned int attempts, double timeout,
+                                         const GroupStateValidityCallbackFn &constraint, const kinematics::KinematicsQueryOptions &options)
+{
```
Above is the location of the `setFromIK` function that @sachinchitta thought might have been removed. Not sure why it appears re-ordered, it is the correct order in the code.

davetcoleman commented:
I have addressed all the concerns of @sachinchitta in https://github.com/davetcoleman/moveit_core/commit/4d5afb5584a405ea94383abc773c5b8b839c65bd, thanks for the feedback
